### PR TITLE
cleanup OpenGLProgram

### PR DESCRIPTION
**Pull request by pixelflinger**
This is the first step before we defer a lot of the glGet* calls.

### CODE REVIEW

**Review by prideout — APPROVED** [review]

File: filament/backend/src/opengl/OpenGLProgram.cpp [hunk 1]
```diff
@@ -35,43 +35,92 @@ using namespace backend;
...
+
+    // check status of program linking and shader compilation, logs error and free all resources
+    // in case of error.
+    // TODO: defer this until first use
+    bool success = OpenGLProgram::checkProgramStatus(name.c_str_safe(),
+            gl.program, gl.shaders);
+
+    // Failing to compile a program can't be fatal, because this will happen a lot in
+    // the material tools. We need to have a better way to handle these errors and
+    // return to the editor. Also note the early "return" statements in this function.
```
Comment (bejado) [inline filament/backend/src/opengl/OpenGLProgram.cpp:65]: Hm, I don't see any `return` statements in `OpenGLProgram::OpenGLProgram`

**Review by bejado — APPROVED** [review]

File: filament/backend/src/opengl/OpenGLProgram.cpp [hunk 2]
```diff
@@ -356,24 +400,6 @@ void logCompilationError(io::ostream& out, Program::Shader shaderType,
     out << "Compilation error in " << to_string(shaderType) << " shader \"" << name << "\":\n"
         << "\"" << error << "\""
         << io::endl;
-
-    size_t lc = 1;
```
Comment (romainguy) [inline filament/backend/src/opengl/OpenGLProgram.cpp:360]: Any reason to remove this? It's quite useful for debugging to see the line numbers and the shader source when we dump an error.

Comment (pixelflinger) [reply filament/backend/src/opengl/OpenGLProgram.cpp:360]: Yes, if we want to keep it, we'll have to keep the source of all programs until we get at least to the first draw.
Even if we do this, then we'll have to free that memory on first draw, but that would be (potentially many) calls to free() -- which could block, etc...

Comment (romainguy) [reply filament/backend/src/opengl/OpenGLProgram.cpp:360]: Could we keep it in debug builds?

File: filament/backend/src/opengl/OpenGLProgram.cpp (hunk 1 above)
Comment (pixelflinger) [reply filament/backend/src/opengl/OpenGLProgram.cpp:65]: good catch.

File: filament/backend/src/opengl/OpenGLProgram.cpp (hunk 2 above)
Comment (pixelflinger) [reply filament/backend/src/opengl/OpenGLProgram.cpp:360]: yeah we could.

Comment (pixelflinger) [reply filament/backend/src/opengl/OpenGLProgram.cpp:360]: done

**Review by romainguy — APPROVED** [review]

**Pull request merged**